zcash_client_sqlite/src/wallet/init/migrations/utxos_to_txos.rs

```diff
@@ -0,0 +1,322 @@
+    }
+
+    fn description(&self) -> &'static str {
+        "Updates transparent UTXO handling to be similar to that for shielded notes."
+    }
+}
+
+impl RusqliteMigration for Migration {
+    type Error = WalletMigrationError;
+
+    fn up(&self, transaction: &rusqlite::Transaction) -> Result<(), WalletMigrationError> {
+        transaction.execute_batch(
+            r#"
+            PRAGMA foreign_keys = OFF;
+            PRAGMA legacy_alter_table = ON;
+
+            CREATE TABLE transactions_new (
+                id_tx INTEGER PRIMARY KEY,
+                txid BLOB NOT NULL UNIQUE,
+                created TEXT,
+                block INTEGER,
+                mined_height INTEGER,
+                tx_index INTEGER,
+                expiry_height INTEGER,
+                raw BLOB,
+                fee INTEGER,
+                FOREIGN KEY (block) REFERENCES blocks(height),
+                CONSTRAINT height_consistency CHECK (block IS NULL OR mined_height = block)
+            );
+
+            -- We may initially set the block height to null, which will mean that the
+            -- transaction may appear to be un-mined until we actually scan the block
+            -- containing the transaction.
+            INSERT INTO transactions_new (txid, block, mined_height)
+            SELECT
+                utxos.prevout_txid,
+                blocks.height,
+                blocks.height
+            FROM utxos
+            LEFT OUTER JOIN blocks ON blocks.height = utxos.height
+            WHERE utxos.prevout_txid NOT IN (
+                SELECT txid FROM transactions
+            );
+
+            INSERT INTO transactions_new
+            SELECT id_tx, txid, created, block, block, tx_index, expiry_height, raw, fee
+            FROM transactions;
+
+            DROP TABLE transactions;
+            ALTER TABLE transactions_new RENAME TO transactions;
+
+            CREATE TABLE transparent_received_outputs (
+                id INTEGER PRIMARY KEY,
+                transaction_id INTEGER NOT NULL,
+                output_index INTEGER NOT NULL,
+                account_id INTEGER NOT NULL,
+                address TEXT NOT NULL,
+                script BLOB NOT NULL,
+                value_zat INTEGER NOT NULL,
+                max_observed_unspent_height INTEGER,
+                FOREIGN KEY (transaction_id) REFERENCES transactions(id_tx),
+                FOREIGN KEY (account_id) REFERENCES accounts(id),
+                CONSTRAINT transparent_output_unique UNIQUE (transaction_id, output_index)
+            );
+
+            INSERT INTO transparent_received_outputs SELECT
+                u.id,
+                t.id_tx,
+                prevout_idx,
+                received_by_account_id,
+                address,
+                script,
+                value_zat,
+                NULL
+            FROM utxos u
+            JOIN transactions t ON t.txid = u.prevout_txid;
```

Comment:
I think if you write this as `LEFT OUTER JOIN`, there will be a constraint error if the transactions table hasn't been updated to include entries for every `prevout_txid` (as should have happened above on lines 55-64), because `t.id_tx` will be NULL and `transaction_id` is declared as `NOT NULL`. That would make sure that if there is a bug in the migration then we can't inadvertently drop any TXOs.